Fix IAD and vis for DS2438 1wire sensor

JumpmanJunior opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
IAD is not the current reading but a switch to enable it.The current reading is in vis. Except from https://owfs.org/uploads/ds2438.html:

> **vis**
> read-only, floating point
> Current sensor reading, as a voltage difference. Value is in volts.
> Actual current depends on Rsens resistor (see datasheet).
> The formula for current is I = vis / Rsens
> with units in Amps and Ohms. 
> 
> **Iad**
> read-write, yes-no
> Current A/D Control flag bit. If "on" (1) current measured at 36.41 Hz. If "off" (0) current measurements won’t be made. See the datasheet for more information. 

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #59103
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

JumpmanJunior committed:
Fix vis and IAD for DS2438

JumpmanJunior committed:
Fix IAD and vis for DS2438

homeassistant commented:
Hi @JumpmanJunior, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#59126).

Once you do that we will be able to review and accept this pull request.

Thanks!

epenet commented:
Would you be able to ajust the const.py file in tests to match the new expected entities?

JumpmanJunior committed:
Fix IAD and vis for DS2438

JumpmanJunior commented:
Done.

garbled1 reviewed (approved):
LGTM

epenet commented on homeassistant/components/onewire/switch.py:
```diff
@@ -58,6 +58,14 @@ class OneWireSwitchEntityDescription(OneWireEntityDescription, SwitchEntityDescr
             for id in DEVICE_KEYS_A_B
         ]
     ),
+    "26": [         
```
Can you make this `tuple` instead of `list`?

epenet commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -184,11 +184,11 @@ class OneWireSensorEntityDescription(OneWireEntityDescription, SensorEntityDescr
             device_class=DEVICE_CLASS_CURRENT,
             entity_registry_enabled_default=False,
-            name="Current",
-            native_unit_of_measurement=ELECTRIC_CURRENT_AMPERE,
```
Maybe just `vis` not "current vis"?

garbled1 replied:
@epenet If the IAD is a switch that tells us to read or not read the vis, isn't that something that would go into the new entity class of CONFIG, ie, so it doesn't show up on the main page, just interior to the integration?

epenet replied:
As pointed out by @garbled1 please add 
```python
entity_category=ENTITY_CATEGORY_CONFIG,
```

garbled1 replied:
to be clear here, you can just add the category like that, as we have 2 other PR's that will be handling the tests for it.  (you inspired me)

epenet commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -184,11 +184,11 @@ class OneWireSensorEntityDescription(OneWireEntityDescription, SensorEntityDescr
         OneWireSensorEntityDescription(
-            key="IAD",
+            key="vis",
             device_class=DEVICE_CLASS_CURRENT,
```
Do we need yo change the device class for vis?

JumpmanJunior committed:
Fix IAD and vis for DS2438

JumpmanJunior committed:
Fix IAD and vis for DS2438

JumpmanJunior committed:
Fix IAD and vis for DS2438

JumpmanJunior commented:
Changed list to tuple.
In the owfs docs vis is described as "Current sensor reading, as a voltage difference". So I think the device class is fine even if the reading is in Volt. But I have renamed it to only "vis".

epenet commented:
@JumpmanJunior can you remove the now obsolete `ELECTRIC_CURRENT_AMPERE` from the imports in `sensor.py`?

epenet commented:
Regarding the device class, please use DEVICE_CLASS_VOLTAGE not DEVICE_CLASS_CURRENT.
(see https://www.home-assistant.io/integrations/sensor/)

epenet commented:
I think @garbled1 is also right about the entity category.
We don't use them yet in `onewire` but it would be good to set it in this case. The tests will need to be adjusted to take it into consideration.

garbled1 commented:
We shouldn't hold up this commit on the entity category thing, especially if nothing else uses it yet in onewire.  I've been actually thinking about some of the more fiddly HB devices and the array of switches on them. (which I ignored previously, because I didn't want to fill the UI up with "turn off this leg of the hub")

For this change, I'm OK with not adding the category, but if you want to take a crack at it, feel free.

epenet commented:
agreed on the entity category.
@JumpmanJunior still needs to fix the device class and the obsolete imports.

And maybe also check the test constants when it is connected through a hub.

JumpmanJunior committed:
Merge branch 'home-assistant:dev' into dev

JumpmanJunior committed:
Fix IAD and vis for DS2438

JumpmanJunior committed:
Fix IAD and vis for DS2438

homeassistant commented:
Hello @JumpmanJunior,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`f216731a407fba5777b9b947b73ea94b35bec19b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;59126&#x2F;commits&#x2F;f216731a407fba5777b9b947b73ea94b35bec19b) - This commit has something that looks like an email address (kuba@lepnet.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

JumpmanJunior committed:
Merge branch 'dev' of github.com:JumpmanJunior/core into dev

homeassistant commented:
Hello @JumpmanJunior,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`be49563384f7e73aa4441b3c159e4501805c4670`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;59126&#x2F;commits&#x2F;be49563384f7e73aa4441b3c159e4501805c4670) - This commit has something that looks like an email address (kuba@lepnet.eu). Maybe try linking that to GitHub?.
- [`f216731a407fba5777b9b947b73ea94b35bec19b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;59126&#x2F;commits&#x2F;f216731a407fba5777b9b947b73ea94b35bec19b) - This commit has something that looks like an email address (kuba@lepnet.eu). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

JumpmanJunior committed:
Merge branch 'home-assistant:dev' into dev

JumpmanJunior commented:
Messed up with my newly set up IDE. But now the requested changes are made.

epenet commented:
@JumpmanJunior do you think you will manage to fix the cla issue?
Or do we need to recreate the PR from scratch?

JumpmanJunior commented:
I already did yesterday before writing my last message. All commits appear as verified.

epenet commented:
This fails on my machine:
`pytest tests/components/onewire --cov=homeassistant/components/onewire --cov-report term-missing`

For `ENTITY_CATEGORY_CONFIG`, you need to make sure that it is imported at the top of `switch.py`:
Replace
`from homeassistant.const import CONF_TYPE`
with
`from homeassistant.const import CONF_TYPE, ENTITY_CATEGORY_CONFIG`

epenet commented on tests/components/onewire/const.py:
```diff
@@ -398,13 +396,22 @@
-                ATTR_INJECT_READS: b"       1",
-                ATTR_STATE: "1.0",
+                ATTR_DEVICE_CLASS: DEVICE_CLASS_VOLTAGE,
+                ATTR_ENTITY_ID: "sensor.26_111111111111_current_vis",
```
Since the name was changed in the definition, it also needs to be adjusted in the test constants:
```suggestion
                ATTR_ENTITY_ID: "sensor.26_111111111111_vis",
```

JumpmanJunior committed:
Fix IAD and vis for DS2438

JumpmanJunior committed:
Merge branch 'dev' of github.com:JumpmanJunior/core into dev

JumpmanJunior commented:
I finally got the tests running also on my machine. I fixed the mentioned and some other issues.

epenet reviewed (approved):
LGTM!
